lib/pos_blockchain.go

```diff
@@ -1832,14 +1834,28 @@ func (bc *Blockchain) getUtxoViewAtBlockHash(blockHash BlockHash) (*UtxoView, er
 		}
 		currentBlock = bc.blockIndexByHash[*currentParentHash]
 		if currentBlock == nil {
-			return nil, errors.Errorf("getUtxoViewAtBlockHash: Block %v not found in block index", blockHash)
+			return nil, errors.Errorf("getUtxoViewAtBlockHash: Block %v not found in block index", currentParentHash)
+		}
+		if currentBlock.IsCommitted() && !currentBlock.Hash.IsEqual(highestCommittedBlock.Hash) {
+			return nil, errors.Errorf(
+				"getUtxoViewAtBlockHash: extends from a committed block that isn't the committed tip")
 		}
 	}
+	if viewAtHash, exists := bc.blockViewCache.Lookup(blockHash); exists {
+		copiedView, err := viewAtHash.(*UtxoView).CopyUtxoView()
+		if err != nil {
+			return nil, errors.Wrapf(err, "getUtxoViewAtBlockHash: Problem copying UtxoView from cache")
+		}
+		return copiedView, nil
+	}
 	// Connect the uncommitted blocks to the tip so that we can validate subsequent blocks
 	utxoView, err := NewUtxoView(bc.db, bc.params, bc.postgres, bc.snapshot, bc.eventManager)
 	if err != nil {
 		return nil, errors.Wrapf(err, "getUtxoViewAtBlockHash: Problem initializing UtxoView")
 	}
+	// TODO: there's another performance enhancement we can make here. If we have a view in the
```

Comment:
Not adding this additional enhancement is fine. We'll only produce a utxoView for a block hash once, and it will be 2-3 blocks past the committed tip. So it's fine if we manually connect all 2-3 blocks when producing it